File: archive/p/python/base64_encode_decode.py
import base64
import binascii
import sys
from typing import NoReturn


def usage() -> NoReturn:
    print("Usage: please provide a mode and a string to encode/decode")
    sys.exit(1)


def base64_encode(s: str) -> str:
    return base64.b64encode(s.encode("ascii")).decode("ascii")


def base64_decode(s: str) -> str:
    return base64.b64decode(s.encode("ascii"), validate=True).decode("ascii")


def main() -> None | NoReturn:
    if len(sys.argv) < 3 or not sys.argv[2]:
        usage()

    mode = sys.argv[1]
    s = sys.argv[2]
    if mode == "encode":
        result = base64_encode(s)
    elif mode == "decode":
        try:
            result = base64_decode(s)
        except binascii.Error:
            usage()
    else:
        usage()

    print(result)


if __name__ == "__main__":
    main()

File: archive/p/php/base64-encode-decode.php
<?php
function usage()
{
    exit("Usage: please provide a mode and a string to encode/decode");
}

// Cannot use the library function directly because it doesn't check if the
// length of the input string is a multiple of 4. The value of false is
// returned if the input string is invalid. Otherwise, the Base64 decoded
// string is returned
function base64_validate_and_decode($str)
{
    // Use the library function if length of input string is a multiple of 4.
    // Otherwise, just return false
    $result = false;
    if (strlen($str) % 4 == 0) {
        $result = base64_decode($str, true);
    }

    return $result;
}

// Exit if arguments are invalid
if (count($argv) < 3 || strlen($argv[2]) < 1)
{
    usage();
}

$mode = $argv[1];
$str = $argv[2];
switch ($mode) {
    case "encode":
        $result = base64_encode($str);
        break;

    case "decode":
        $result = base64_validate_and_decode($str);
        if ($result === false) {
            usage();
        }
        break;

    default:
        usage();
}

echo $result;
